![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2FApplied+ML%2FSolution+Prototypes%2Fproduct-hierarchy&file=product-hierarchy.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2FApplied%2520ML%2FSolution%2520Prototypes%2Fproduct-hierarchy%2Fproduct-hierarchy.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>      
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b> 
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520ML/Solution%2520Prototypes/product-hierarchy/product-hierarchy.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a> 
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520ML/Solution%2520Prototypes/product-hierarchy/product-hierarchy.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a> 
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520ML/Solution%2520Prototypes/product-hierarchy/product-hierarchy.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520ML/Solution%2520Prototypes/product-hierarchy/product-hierarchy.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a> 
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b> 
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a> 
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy.ipynb"><img src="https://www.svgrepo.com/download/5445/download-button.svg" alt="Download icon" width="20px"></a> <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/Applied%20ML/Solution%20Prototypes/product-hierarchy/product-hierarchy.ipynb">Download Notebook</a> <i>(right-click and "Save As")</i>
  </td>
</tr>
</table>

# Product Hierarchy Classification

When a retailer manages thousands (or millions) of SKUs, every product lives in a **hierarchy** — typically Department → Category, sometimes deeper. New products arrive constantly from vendors, acquisitions, and marketplace sellers, often with inconsistent or minimal descriptions. Placing them correctly matters: it drives search, recommendations, reporting, and planogram layout.

**Why is this hard?**

Consider these incoming vendor descriptions:

| Vendor Description | Correct Placement | Why It's Tricky |
|---|---|---|
| *"65-inch flat screen display panel for home theater"* | Electronics → Televisions | Never says "TV" or "television" |
| *"durable blue denim trousers for guys"* | Mens → Clothing | Says "trousers" not "jeans", "guys" not "mens" |
| *"cordless stick vacuum with detangling motorbar"* | Home & Kitchen → Cleaning | Could be "Small Appliances" — depends on your taxonomy |
| *"waterproof digital watch for kids"* | Children → Accessories | Could be Electronics → Wearables in a different hierarchy |

Simple keyword rules break down fast. You need approaches that understand **meaning**, not just matching characters.

**This notebook explores three classification paradigms across nine approaches:**

1. **Flat Classification** — Treat every Department/Category path as a unique label. Simple to build, but ignores the tree structure. *(Approaches 1-4: Fuzzy Matching, Semantic Embeddings, Centroid Matching, Gradient Boosting)*

2. **Local Classifiers (Top-Down)** — Classify the Department first, then route to a specialized Category classifier within that department. Leverages hierarchy, but cascading errors at the top level are fatal. *(Approaches 5-7: Hierarchical Embeddings, Hierarchical Centroids, Hierarchical GBC)*

3. **Multi-Task Learning** — A single neural network with shared hidden layers and separate output heads for Department and Category. Learns hierarchical relationships naturally without cascading errors. *(Approaches 8-9: Independent Heads, Conditional Category Head)*

We generate a synthetic product catalog using **Gemini structured output** with Pydantic schemas, then run all approaches and compare results.

---
## Setup

In [1]:
PROJECT_ID = 'statmike-mlops-349915'  # <-- Replace with your project ID
REGION = 'us-central1'

### Environment

> **Already set up the project environment?** The cell below is a no-op — packages are already in your kernel. See the [README](README.md) for uv setup details.
>
> **Running standalone** (Colab, Colab Enterprise, Vertex AI Workbench)? The cell below installs required packages into your current kernel.

In [2]:
import subprocess, sys, shutil

def install(*packages):
    uv = shutil.which('uv')
    if uv:
        subprocess.check_call([uv, 'pip', 'install', '-q', '--python', sys.executable, *packages])
    else:
        subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', '--upgrade', *packages])

install('google-genai', 'thefuzz', 'scikit-learn', 'pandas', 'numpy', 'pydantic', 'keras', 'tensorflow')

In [3]:
# Authenticate (Colab only — other environments use Application Default Credentials)
try:
    from google.colab import auth
    auth.authenticate_user()
except ImportError:
    pass  # Not in Colab — ADC is used automatically

In [6]:
import os

os.environ['GOOGLE_GENAI_USE_VERTEXAI'] = 'True'
os.environ['GOOGLE_CLOUD_PROJECT'] = PROJECT_ID
os.environ['GOOGLE_CLOUD_LOCATION'] = REGION

from google import genai
from google.genai import types

client = genai.Client()
GEMINI_MODEL = 'gemini-2.5-flash'
EMBEDDING_MODEL = 'gemini-embedding-001'

print(f'GenAI client ready — project: {PROJECT_ID}, region: {REGION}')

GenAI client ready — project: statmike-mlops-349915, region: us-central1


---
## Define the Product Hierarchy

Our taxonomy has **5 departments**, each with **3-5 categories**. This is a simplified version of what a real retailer might use — enough to demonstrate meaningful classification challenges without being overwhelming.

In practice, hierarchies are often stored in a database or config file. Here we define it as a simple dictionary.

In [7]:
HIERARCHY = {
    'Mens': ['Clothing', 'Shoes', 'Accessories', 'Outerwear'],
    'Womens': ['Clothing', 'Shoes', 'Accessories', 'Outerwear', 'Activewear'],
    'Children': ['Clothing', 'Shoes', 'Toys', 'School Supplies'],
    'Electronics': ['Mobile Phones', 'Televisions', 'Audio', 'Computers'],
    'Home & Kitchen': ['Small Appliances', 'Cookware', 'Furniture', 'Cleaning'],
}

# Flat list of all valid (department, category) pairs
ALL_DEPARTMENTS = list(HIERARCHY.keys())
ALL_CATEGORIES = sorted(set(cat for cats in HIERARCHY.values() for cat in cats))

print(f'{len(HIERARCHY)} departments, {sum(len(v) for v in HIERARCHY.values())} total categories:\n')
for dept, categories in HIERARCHY.items():
    print(f'  {dept}')
    for cat in categories:
        print(f'    - {cat}')

5 departments, 21 total categories:

  Mens
    - Clothing
    - Shoes
    - Accessories
    - Outerwear
  Womens
    - Clothing
    - Shoes
    - Accessories
    - Outerwear
    - Activewear
  Children
    - Clothing
    - Shoes
    - Toys
    - School Supplies
  Electronics
    - Mobile Phones
    - Televisions
    - Audio
    - Computers
  Home & Kitchen
    - Small Appliances
    - Cookware
    - Furniture
    - Cleaning


---
## Generate Synthetic Catalog with Structured Output

Instead of hard-coding product data, we'll use **Gemini structured output** to generate a realistic catalog. We define [Pydantic](https://docs.pydantic.dev/) models that describe the exact shape of the data we want, including `Literal` type constraints that restrict `department` and `category` to valid values from our hierarchy. Gemini's structured output guarantees the response conforms to this schema — no post-processing or validation needed.

This pattern is powerful beyond this demo: in production, you could use it to parse unstructured vendor PDFs, normalize messy spreadsheet imports, or transform free-text product submissions into structured catalog entries.

In [8]:
from pydantic import BaseModel, Field
from typing import Literal
import pandas as pd
import numpy as np

# Build Literal types dynamically from our hierarchy
DepartmentType = Literal['Mens', 'Womens', 'Children', 'Electronics', 'Home & Kitchen']
CategoryType = Literal[
    'Clothing', 'Shoes', 'Accessories', 'Outerwear', 'Activewear',
    'Toys', 'School Supplies',
    'Mobile Phones', 'Televisions', 'Audio', 'Computers',
    'Small Appliances', 'Cookware', 'Furniture', 'Cleaning',
]


class CatalogProduct(BaseModel):
    """A product in the existing catalog with known department and category."""
    sku: str = Field(description='Unique product identifier, e.g. SKU-1001')
    brand: str = Field(description='Brand name, use real well-known brands')
    name: str = Field(description='Short product name, 2-5 words')
    description: str = Field(description='Product description, 10-20 words, realistic retail language')
    department: DepartmentType
    category: CategoryType
    retail_price: float = Field(description='Retail price in USD')


class ProductCatalog(BaseModel):
    """A collection of catalog products."""
    products: list[CatalogProduct]


print('Pydantic schemas defined.')
print(f'CatalogProduct fields: {list(CatalogProduct.model_fields.keys())}')

Pydantic schemas defined.
CatalogProduct fields: ['sku', 'brand', 'name', 'description', 'department', 'category', 'retail_price']


In [9]:
# Build a prompt that describes the hierarchy and asks for diverse products
hierarchy_description = '\n'.join(
    f'  {dept}: {", ".join(cats)}' for dept, cats in HIERARCHY.items()
)

catalog_prompt = f"""Generate exactly 40 realistic retail products for an existing product catalog.

Product hierarchy (Department: Categories):
{hierarchy_description}

Requirements:
- Distribute products across ALL departments and categories (at least 2 products per category)
- Use real, well-known brand names appropriate to each category
- Write realistic product descriptions (10-20 words) using typical retail language
- Use realistic retail prices in USD
- SKUs should be formatted as SKU-1001, SKU-1002, etc.
- Each product's category MUST be valid for its department (e.g. "Toys" only under "Children")
"""

catalog_response = client.models.generate_content(
    model=GEMINI_MODEL,
    contents=catalog_prompt,
    config=types.GenerateContentConfig(
        response_mime_type='application/json',
        response_schema=ProductCatalog,
    ),
)

catalog = ProductCatalog.model_validate_json(catalog_response.text)
catalog_df = pd.DataFrame([p.model_dump() for p in catalog.products])

print(f'Generated {len(catalog_df)} catalog products across {catalog_df["department"].nunique()} departments and {catalog_df["category"].nunique()} categories')
print(f'\nProducts per department:')
print(catalog_df['department'].value_counts().to_string())

Generated 40 catalog products across 5 departments and 15 categories

Products per department:
department
Womens            10
Mens               8
Children           8
Electronics        8
Home & Kitchen     6


In [10]:
# Review the generated catalog
catalog_df

,sku,brand,name,description,department,category,retail_price
0,SKU-1001,Levi's,501 Original Jeans,Classic straight-leg denim for everyday comfor...,Mens,Clothing,69.99
1,SKU-1002,Ralph Lauren,Polo Oxford Shirt,Iconic button-down shirt with signature pony e...,Mens,Clothing,98.00
2,SKU-1003,Nike,Air Force 1 Sneakers,"Legendary basketball shoe, reimagined for stre...",Mens,Shoes,110.00
3,SKU-1004,Timberland,6-Inch Premium Boots,Waterproof leather boots for rugged outdoor ad...,Mens,Shoes,199.99
4,SKU-1005,Fossil,Leather Bi-fold Wallet,Stylish and durable genuine leather wallet wit...,Mens,Accessories,45.00
5,SKU-1006,Ray-Ban,Wayfarer Sunglasses,Iconic sunglasses offering timeless design and...,Mens,Accessories,165.00
6,SKU-1007,The North Face,Denali Fleece Jacket,"Warm, durable fleece jacket perfect for layeri...",Mens,Outerwear,179.00
7,SKU-1008,Columbia,Waterproof Rain Jacket,"Lightweight, breathable jacket provides reliab...",Mens,Outerwear,90.00
8,SKU-1009,H&M,Ribbed Knit Sweater,"Soft, comfortable ribbed sweater, perfect for ...",Womens,Clothing,29.99
9,SKU-1010,Zara,High-Waist Skinny Jeans,Flattering high-waisted jeans with a comfortab...,Womens,Clothing,49.90


---
## Generate New Products to Classify

Now we simulate **incoming vendor products** — these arrive without department or category labels. The vendor descriptions are intentionally varied: they may use synonyms, informal language, or ambiguous terms that don't map cleanly to our taxonomy.

We use a separate Pydantic schema (no `department` or `category` fields) and prompt Gemini to generate products that are deliberately challenging to classify.

In [11]:
class NewProduct(BaseModel):
    """An incoming vendor product without department/category labels."""
    sku: str = Field(description='Unique product identifier, e.g. NEW-9001')
    brand: str = Field(description='Brand name')
    name: str = Field(description='Short product name, 2-5 words')
    description: str = Field(description='Product description, 10-20 words, may use informal language or synonyms')
    cost: float = Field(description='Wholesale cost in USD')
    retail_price: float = Field(description='Suggested retail price in USD')


class NewProductBatch(BaseModel):
    """A batch of new products to classify."""
    products: list[NewProduct]


new_products_prompt = f"""Generate exactly 8 new vendor products that need to be classified into a product hierarchy.

The hierarchy they will be classified into:
{hierarchy_description}

Requirements:
- Do NOT include department or category fields — these are unknown and need to be predicted
- Make the descriptions deliberately challenging to classify:
  - Use synonyms instead of obvious category words (e.g. "trousers" instead of "jeans", "display panel" instead of "TV")
  - Include ambiguous items that could fit multiple categories (e.g. a smart watch could be Electronics or Accessories)
  - Use informal or vendor-specific language (e.g. "kicks" instead of "shoes")
  - Include at least one item that spans departments (e.g. a kids' electronic toy)
- Use real brand names
- SKUs should be formatted as NEW-9001, NEW-9002, etc.
- Include realistic cost and retail prices
"""

new_response = client.models.generate_content(
    model=GEMINI_MODEL,
    contents=new_products_prompt,
    config=types.GenerateContentConfig(
        response_mime_type='application/json',
        response_schema=NewProductBatch,
    ),
)

new_batch = NewProductBatch.model_validate_json(new_response.text)
new_df = pd.DataFrame([p.model_dump() for p in new_batch.products])

print(f'Generated {len(new_df)} new products to classify:\n')
for _, row in new_df.iterrows():
    print(f'  {row["sku"]} | {row["brand"]} {row["name"]}')
    print(f'         "{row["description"]}"\n')

Generated 8 new products to classify:

  NEW-9001 | Levi's Classic Indigo Legwear
         "Durable denim trousers, perfect for everyday wear and robust activities. Features multiple pockets."

  NEW-9002 | Steve Madden Glimmering Stilettos
         "High-heeled footwear with a metallic finish, ideal for evening events and special occasions."

  NEW-9003 | Samsung Quantum View Panel
         "Ultra-HD large format visual display unit for an immersive viewing experience. Smart features included."

  NEW-9004 | Keurig Rapid Brew System
         "Single-serve hot beverage preparation appliance. Quick, efficient, and fits perfectly on any countertop."

  NEW-9005 | Fitbit Aura Fitness Tracker
         "Wrist-worn intelligent device monitoring activity levels and notifications. Compatible with various platforms."

  NEW-9006 | VTech Dino Code-bot
         "Interactive robotic plaything designed to teach coding basics through engaging games for young learners."

  NEW-9007 | The North Face S

In [12]:
# Review the new products
new_df

,sku,brand,name,description,cost,retail_price
0,NEW-9001,Levi's,Classic Indigo Legwear,"Durable denim trousers, perfect for everyday w...",35.0,70.0
1,NEW-9002,Steve Madden,Glimmering Stilettos,"High-heeled footwear with a metallic finish, i...",45.0,90.0
2,NEW-9003,Samsung,Quantum View Panel,Ultra-HD large format visual display unit for ...,750.0,1500.0
3,NEW-9004,Keurig,Rapid Brew System,Single-serve hot beverage preparation applianc...,60.0,120.0
4,NEW-9005,Fitbit,Aura Fitness Tracker,Wrist-worn intelligent device monitoring activ...,110.0,220.0
5,NEW-9006,VTech,Dino Code-bot,Interactive robotic plaything designed to teac...,40.0,80.0
6,NEW-9007,The North Face,Summit Trail Jacket,"Lightweight, breathable external garment offer...",90.0,180.0
7,NEW-9008,Lululemon,Align Sculpt Tights,"High-waisted, flexible performance stretch bot...",55.0,110.0


---
# Part 1: Flat Classification

In flat classification, we treat every **Department / Category** combination as a unique, independent label. A model trained this way doesn't "know" that `Mens / Clothing` and `Womens / Clothing` share anything in common — they're as different as `Electronics / Televisions`. This is the simplest approach to implement and maintain, and it works well when the taxonomy is shallow (2-3 levels) and the number of leaf categories is manageable.

**The drawback:** If the model predicts `Mens / Televisions` instead of `Electronics / Televisions`, it gets penalized the same as predicting `Home & Kitchen / Cookware`. There's no notion of "close but wrong department."

## Feature Engineering

Before comparing approaches, we combine key product fields into a single text string. This gives each approach the same input — a unified "search text" that captures brand, name, and description together.

In [13]:
def create_text_blob(row):
    """Combine key product attributes into a single descriptive string."""
    return f"Brand: {row.get('brand', '')} | Name: {row.get('name', '')} | Desc: {row.get('description', '')}"

catalog_df['search_text'] = catalog_df.apply(create_text_blob, axis=1)
new_df['search_text'] = new_df.apply(create_text_blob, axis=1)

# Show an example
print('Example search_text:')
print(f'  Catalog: {catalog_df["search_text"].iloc[0]}')
print(f'  New:     {new_df["search_text"].iloc[0]}')

Example search_text:
  Catalog: Brand: Levi's | Name: 501 Original Jeans | Desc: Classic straight-leg denim for everyday comfort and timeless style. Durable, versatile wardrobe essential.
  New:     Brand: Levi's | Name: Classic Indigo Legwear | Desc: Durable denim trousers, perfect for everyday wear and robust activities. Features multiple pockets.


---
## Approach 1: Fuzzy String Matching (Levenshtein Distance)

The simplest approach: compare the new product's text against every catalog item using **character-level edit distance**. We use `thefuzz`'s `token_set_ratio`, which:

- Tokenizes both strings into words
- Finds the intersection (shared tokens) and remainder
- Computes a ratio that's invariant to word order

**Strengths:** Fast, no API calls, no training needed, works offline.

**Weaknesses:** Purely syntactic — "trousers" and "jeans" look nothing alike character-by-character, even though they mean the same thing. Similarly, "display panel" won't match well against "television."

In [15]:
from thefuzz import fuzz

def classify_with_levenshtein(new_text, catalog_df):
    """Find the best matching catalog product using fuzzy string matching."""
    best_score = -1
    best_match_idx = -1

    for idx, row in catalog_df.iterrows():
        score = fuzz.token_set_ratio(new_text, row['search_text'])
        if score > best_score:
            best_score = score
            best_match_idx = idx

    best_match = catalog_df.loc[best_match_idx]
    return best_match['department'], best_match['category'], best_score, best_match['name']

print('--- APPROACH 1: FUZZY STRING MATCHING (LEVENSHTEIN) ---\n')

levenshtein_results = []
for _, new_prod in new_df.iterrows():
    dept, cat, score, matched_name = classify_with_levenshtein(new_prod['search_text'], catalog_df)
    levenshtein_results.append({'sku': new_prod['sku'], 'department': dept, 'category': cat, 'score': score})
    print(f'{new_prod["name"]}')
    print(f'  -> {dept} / {cat}')
    print(f'  -> Matched: "{matched_name}" (score: {score}/100)\n')

--- APPROACH 1: FUZZY STRING MATCHING (LEVENSHTEIN) ---

Classic Indigo Legwear
  -> Mens / Clothing
  -> Matched: "501 Original Jeans" (score: 65/100)

Glimmering Stilettos
  -> Children / School Supplies
  -> Matched: "SuperBreak Backpack" (score: 56/100)

Quantum View Panel
  -> Electronics / Mobile Phones
  -> Matched: "Galaxy S24" (score: 58/100)

Rapid Brew System
  -> Home & Kitchen / Small Appliances
  -> Matched: "K-Classic Coffee Maker" (score: 58/100)

Aura Fitness Tracker
  -> Mens / Outerwear
  -> Matched: "Waterproof Rain Jacket" (score: 54/100)

Dino Code-bot
  -> Womens / Accessories
  -> Matched: "Charm Bracelet" (score: 55/100)

Summit Trail Jacket
  -> Mens / Outerwear
  -> Matched: "Waterproof Rain Jacket" (score: 66/100)

Align Sculpt Tights
  -> Womens / Activewear
  -> Matched: "Align Leggings" (score: 61/100)



---
## Approach 2: Semantic Embeddings (Nearest Neighbor)

Embeddings convert text into high-dimensional vectors where **semantically similar text is close together**. Unlike Levenshtein, embeddings understand that "trousers" ≈ "jeans" and "display panel" ≈ "television" because they appear in similar contexts in training data.

We use Google's `gemini-embedding-001` model with the **`CLASSIFICATION`** task type, which optimizes the embedding vectors for distinguishing between categories — exactly what we need for product hierarchy classification.

For each new product, we find the catalog item with the highest **cosine similarity** (ranges from -1 to 1, where 1 = identical meaning) and inherit its department and category.

In [17]:
from sklearn.metrics.pairwise import cosine_similarity

# Step 1: Embed the entire catalog (done once, could be cached in production)
print('Embedding catalog products...')
catalog_embed_response = client.models.embed_content(
    model=EMBEDDING_MODEL,
    contents=catalog_df['search_text'].tolist(),
    config=types.EmbedContentConfig(task_type='CLASSIFICATION'),
)
catalog_embeddings = np.array([emb.values for emb in catalog_embed_response.embeddings])
print(f'  {catalog_embeddings.shape[0]} products embedded into {catalog_embeddings.shape[1]}-dimensional vectors')

Embedding catalog products...
  40 products embedded into 3072-dimensional vectors


In [18]:
def classify_with_embeddings(new_text, catalog_df, catalog_vecs):
    """Find the most semantically similar catalog product using embeddings."""
    response = client.models.embed_content(
        model=EMBEDDING_MODEL,
        contents=new_text,
        config=types.EmbedContentConfig(task_type='CLASSIFICATION'),
    )
    new_vec = np.array([response.embeddings[0].values])
    similarities = cosine_similarity(new_vec, catalog_vecs)[0]
    best_idx = np.argmax(similarities)
    best_match = catalog_df.iloc[best_idx]
    return best_match['department'], best_match['category'], similarities[best_idx], best_match['name']

print('--- APPROACH 2: SEMANTIC EMBEDDINGS (NEAREST NEIGHBOR) ---\n')

embedding_results = []
for _, new_prod in new_df.iterrows():
    dept, cat, score, matched_name = classify_with_embeddings(
        new_prod['search_text'], catalog_df, catalog_embeddings,
    )
    embedding_results.append({'sku': new_prod['sku'], 'department': dept, 'category': cat, 'score': round(score, 3)})
    print(f'{new_prod["name"]}')
    print(f'  -> {dept} / {cat}')
    print(f'  -> Matched: "{matched_name}" (cosine similarity: {score:.3f})\n')

--- APPROACH 2: SEMANTIC EMBEDDINGS (NEAREST NEIGHBOR) ---

Classic Indigo Legwear
  -> Mens / Clothing
  -> Matched: "501 Original Jeans" (cosine similarity: 0.920)

Glimmering Stilettos
  -> Womens / Shoes
  -> Matched: "Arizona Sandals" (cosine similarity: 0.847)

Quantum View Panel
  -> Electronics / Televisions
  -> Matched: "BRAVIA XR 4K TV" (cosine similarity: 0.873)

Rapid Brew System
  -> Home & Kitchen / Small Appliances
  -> Matched: "K-Classic Coffee Maker" (cosine similarity: 0.896)

Aura Fitness Tracker
  -> Womens / Accessories
  -> Matched: "Charm Bracelet" (cosine similarity: 0.834)

Dino Code-bot
  -> Children / Toys
  -> Matched: "Wooden Play Kitchen" (cosine similarity: 0.830)

Summit Trail Jacket
  -> Womens / Outerwear
  -> Matched: "Down Sweater Jacket" (cosine similarity: 0.922)

Align Sculpt Tights
  -> Womens / Activewear
  -> Matched: "Align Leggings" (cosine similarity: 0.916)



---
## Approach 3: Centroid Matching (Category Averages)

Instead of matching against individual products, we compute the **centroid** (mean embedding vector) for each Department/Category pair. This represents the "average" product in that category.

**Why this can be better than nearest neighbor:**
- A single unusual product won't skew the result — the centroid smooths out outliers
- With large catalogs, it's much faster — compare against ~20 centroids instead of thousands of products
- It captures the "essence" of a category rather than depending on one lucky match

**Trade-off:** If a category has very diverse products (e.g., "Accessories" includes belts, watches, and wallets), the centroid may be a vague average that doesn't represent any real product well.

In [19]:
# Compute centroids: average embedding per (department, category) pair
category_centroids = {}
for (dept, cat), group in catalog_df.groupby(['department', 'category']):
    group_vectors = catalog_embeddings[group.index.tolist()]
    category_centroids[(dept, cat)] = np.mean(group_vectors, axis=0)

print(f'Computed {len(category_centroids)} category centroids:')
for (dept, cat) in sorted(category_centroids.keys()):
    count = len(catalog_df[(catalog_df['department'] == dept) & (catalog_df['category'] == cat)])
    print(f'  {dept} / {cat} ({count} products)')

Computed 21 category centroids:
  Children / Clothing (2 products)
  Children / School Supplies (2 products)
  Children / Shoes (2 products)
  Children / Toys (2 products)
  Electronics / Audio (2 products)
  Electronics / Computers (2 products)
  Electronics / Mobile Phones (2 products)
  Electronics / Televisions (2 products)
  Home & Kitchen / Cleaning (1 products)
  Home & Kitchen / Cookware (2 products)
  Home & Kitchen / Furniture (1 products)
  Home & Kitchen / Small Appliances (2 products)
  Mens / Accessories (2 products)
  Mens / Clothing (2 products)
  Mens / Outerwear (2 products)
  Mens / Shoes (2 products)
  Womens / Accessories (2 products)
  Womens / Activewear (2 products)
  Womens / Clothing (2 products)
  Womens / Outerwear (2 products)
  Womens / Shoes (2 products)


In [20]:
def classify_with_centroids(new_text, centroids_dict):
    """Classify by comparing against category centroid vectors."""
    response = client.models.embed_content(
        model=EMBEDDING_MODEL,
        contents=new_text,
        config=types.EmbedContentConfig(task_type='CLASSIFICATION'),
    )
    new_vec = np.array([response.embeddings[0].values])

    best_score = -1
    best_dept, best_cat = None, None
    for (dept, cat), centroid_vec in centroids_dict.items():
        score = cosine_similarity(new_vec, [centroid_vec])[0][0]
        if score > best_score:
            best_score = score
            best_dept, best_cat = dept, cat

    return best_dept, best_cat, best_score

print('--- APPROACH 3: CENTROID MATCHING (CATEGORY AVERAGES) ---\n')

centroid_results = []
for _, new_prod in new_df.iterrows():
    dept, cat, score = classify_with_centroids(new_prod['search_text'], category_centroids)
    centroid_results.append({'sku': new_prod['sku'], 'department': dept, 'category': cat, 'score': round(score, 3)})
    print(f'{new_prod["name"]}')
    print(f'  -> {dept} / {cat}')
    print(f'  -> Cosine similarity to centroid: {score:.3f}\n')

--- APPROACH 3: CENTROID MATCHING (CATEGORY AVERAGES) ---

Classic Indigo Legwear
  -> Mens / Clothing
  -> Cosine similarity to centroid: 0.918

Glimmering Stilettos
  -> Mens / Shoes
  -> Cosine similarity to centroid: 0.872

Quantum View Panel
  -> Electronics / Televisions
  -> Cosine similarity to centroid: 0.886

Rapid Brew System
  -> Home & Kitchen / Small Appliances
  -> Cosine similarity to centroid: 0.913

Aura Fitness Tracker
  -> Womens / Activewear
  -> Cosine similarity to centroid: 0.835

Dino Code-bot
  -> Children / Toys
  -> Cosine similarity to centroid: 0.852

Summit Trail Jacket
  -> Mens / Outerwear
  -> Cosine similarity to centroid: 0.927

Align Sculpt Tights
  -> Womens / Activewear
  -> Cosine similarity to centroid: 0.931



---
## Approach 4: Gradient Boosting Classifier (Supervised ML)

Here we train a **supervised machine learning model** on the catalog embeddings. The embedding vectors become features, and the Department/Category pair becomes the label. The classifier learns decision boundaries in embedding space — potentially capturing non-linear patterns that simple cosine similarity might miss.

We use scikit-learn's `GradientBoostingClassifier` for simplicity. In production with larger datasets, you'd likely use XGBoost, LightGBM, or even a neural network.

**Note:** With only ~40 training examples across ~20 classes, this classifier is severely data-limited. It will likely perform comparably to (or worse than) the embedding approaches above. The real value of this approach emerges with hundreds or thousands of labeled products per category.

In [21]:
from sklearn.ensemble import GradientBoostingClassifier

# Combine department and category into a single label for the classifier
y_train = (catalog_df['department'] + ' / ' + catalog_df['category']).values
X_train = catalog_embeddings

print('Training Gradient Boosting Classifier...')
clf = GradientBoostingClassifier(n_estimators=50, learning_rate=0.1, max_depth=3, random_state=42)
clf.fit(X_train, y_train)
print(f'  Trained on {len(X_train)} samples across {len(clf.classes_)} classes')
print(f'  Classes: {list(clf.classes_)}')

Training Gradient Boosting Classifier...


/home/user/git/vertex-ai-mlops/Applied ML/Solution Prototypes/product-hierarchy/.venv/lib/python3.13/site-packages/sklearn/ensemble/_gb.py:1524: UserWarning: The number of unique classes is greater than 50% of the number of samples. `y` could represent a regression problem, not a classification problem.
  check_classification_targets(y)


  Trained on 40 samples across 21 classes
  Classes: ['Children / Clothing', 'Children / School Supplies', 'Children / Shoes', 'Children / Toys', 'Electronics / Audio', 'Electronics / Computers', 'Electronics / Mobile Phones', 'Electronics / Televisions', 'Home & Kitchen / Cleaning', 'Home & Kitchen / Cookware', 'Home & Kitchen / Furniture', 'Home & Kitchen / Small Appliances', 'Mens / Accessories', 'Mens / Clothing', 'Mens / Outerwear', 'Mens / Shoes', 'Womens / Accessories', 'Womens / Activewear', 'Womens / Clothing', 'Womens / Outerwear', 'Womens / Shoes']


In [22]:
def classify_with_gbc(new_text, classifier):
    """Classify using the trained Gradient Boosting model."""
    response = client.models.embed_content(
        model=EMBEDDING_MODEL,
        contents=new_text,
        config=types.EmbedContentConfig(task_type='CLASSIFICATION'),
    )
    new_vec = np.array([response.embeddings[0].values])
    pred_label = classifier.predict(new_vec)[0]
    confidence = np.max(classifier.predict_proba(new_vec))
    dept, cat = pred_label.split(' / ')
    return dept, cat, confidence

print('--- APPROACH 4: GRADIENT BOOSTING CLASSIFIER ---\n')

gbc_results = []
for _, new_prod in new_df.iterrows():
    dept, cat, conf = classify_with_gbc(new_prod['search_text'], clf)
    gbc_results.append({'sku': new_prod['sku'], 'department': dept, 'category': cat, 'score': round(conf, 3)})
    print(f'{new_prod["name"]}')
    print(f'  -> {dept} / {cat}')
    print(f'  -> Confidence: {conf:.3f}\n')

--- APPROACH 4: GRADIENT BOOSTING CLASSIFIER ---

Classic Indigo Legwear
  -> Womens / Clothing
  -> Confidence: 0.149

Glimmering Stilettos
  -> Womens / Shoes
  -> Confidence: 0.488

Quantum View Panel
  -> Children / School Supplies
  -> Confidence: 0.731

Rapid Brew System
  -> Home & Kitchen / Small Appliances
  -> Confidence: 0.512

Aura Fitness Tracker
  -> Children / Shoes
  -> Confidence: 0.657

Dino Code-bot
  -> Children / Toys
  -> Confidence: 0.875

Summit Trail Jacket
  -> Mens / Outerwear
  -> Confidence: 0.741

Align Sculpt Tights
  -> Womens / Activewear
  -> Confidence: 0.999



---
## Flat Classification: Compare All Approaches

Let's put all four flat approaches side-by-side. When all four converge on the same classification, we can have high confidence. When they diverge, it highlights genuinely ambiguous products.

In [23]:
# Build a comparison DataFrame
comparison_rows = []
for i, (_, new_prod) in enumerate(new_df.iterrows()):
    lev = levenshtein_results[i]
    emb = embedding_results[i]
    cen = centroid_results[i]
    gbc = gbc_results[i]

    row = {
        'Product': f'{new_prod["brand"]} {new_prod["name"]}',
        'Levenshtein': f'{lev["department"]} / {lev["category"]}',
        'Embeddings': f'{emb["department"]} / {emb["category"]}',
        'Centroid': f'{cen["department"]} / {cen["category"]}',
        'GBC': f'{gbc["department"]} / {gbc["category"]}',
    }

    # Check agreement
    predictions = {row['Levenshtein'], row['Embeddings'], row['Centroid'], row['GBC']}
    row['Agreement'] = f'{4 - len(predictions) + 1}/4'

    comparison_rows.append(row)

comparison_df = pd.DataFrame(comparison_rows)
comparison_df

,Product,Levenshtein,Embeddings,Centroid,GBC,Agreement
0,Levi's Classic Indigo Legwear,Mens / Clothing,Mens / Clothing,Mens / Clothing,Womens / Clothing,3/4
1,Steve Madden Glimmering Stilettos,Children / School Supplies,Womens / Shoes,Mens / Shoes,Womens / Shoes,2/4
2,Samsung Quantum View Panel,Electronics / Mobile Phones,Electronics / Televisions,Electronics / Televisions,Children / School Supplies,2/4
3,Keurig Rapid Brew System,Home & Kitchen / Small Appliances,Home & Kitchen / Small Appliances,Home & Kitchen / Small Appliances,Home & Kitchen / Small Appliances,4/4
4,Fitbit Aura Fitness Tracker,Mens / Outerwear,Womens / Accessories,Womens / Activewear,Children / Shoes,1/4
5,VTech Dino Code-bot,Womens / Accessories,Children / Toys,Children / Toys,Children / Toys,3/4
6,The North Face Summit Trail Jacket,Mens / Outerwear,Womens / Outerwear,Mens / Outerwear,Mens / Outerwear,3/4
7,Lululemon Align Sculpt Tights,Womens / Activewear,Womens / Activewear,Womens / Activewear,Womens / Activewear,4/4


In [24]:
# Score summary: compare confidence/similarity scores across approaches
score_rows = []
for i, (_, new_prod) in enumerate(new_df.iterrows()):
    score_rows.append({
        'Product': f'{new_prod["brand"]} {new_prod["name"]}',
        'Levenshtein (0-100)': levenshtein_results[i]['score'],
        'Embeddings (cos sim)': embedding_results[i]['score'],
        'Centroid (cos sim)': centroid_results[i]['score'],
        'GBC (confidence)': gbc_results[i]['score'],
    })

score_df = pd.DataFrame(score_rows)
score_df

,Product,Levenshtein (0-100),Embeddings (cos sim),Centroid (cos sim),GBC (confidence)
0,Levi's Classic Indigo Legwear,65,0.920,0.918,0.149
1,Steve Madden Glimmering Stilettos,56,0.847,0.872,0.488
2,Samsung Quantum View Panel,58,0.873,0.886,0.731
3,Keurig Rapid Brew System,58,0.896,0.913,0.512
4,Fitbit Aura Fitness Tracker,54,0.834,0.835,0.657
5,VTech Dino Code-bot,55,0.830,0.852,0.875
6,The North Face Summit Trail Jacket,66,0.922,0.927,0.741
7,Lululemon Align Sculpt Tights,61,0.916,0.931,0.999


### Flat Classification: Approach Summary

| Approach | Strengths | Weaknesses | Best For |
|----------|-----------|------------|----------|
| **Fuzzy Matching** | Fast, no API calls, no training | Purely syntactic — misses synonyms and semantic meaning | Quick baseline, exact/near-exact matches |
| **Embeddings (NN)** | Understands meaning, handles synonyms | Depends on individual product quality, API calls needed | Small-to-medium catalogs, good descriptions |
| **Centroid Matching** | Robust to outliers, fast at scale | Diverse categories produce vague centroids | Large catalogs, well-defined categories |
| **Gradient Boosting** | Learns non-linear boundaries, confidence scores | Needs substantial labeled data to train well | Large labeled datasets, complex taxonomies |

**Key limitation of all flat approaches:** They treat `Department / Category` as a single opaque label. The hierarchy structure — that "Clothing" under "Mens" and "Clothing" under "Womens" are related — is completely ignored. Next, we'll exploit that structure.

---
# Part 2: Local Classifiers (Top-Down / Hierarchical)

Local classification takes a fundamentally different approach: instead of one model that predicts the full path, you **classify iteratively down the tree**. First predict the Department, then — within that department — predict the Category.

**When to use local classifiers over flat:**

| Factor | Flat Classification | Local Classifiers |
|--------|--------------------|--------------------|
| **Hierarchy depth** | Works well at 1-2 levels | Essential at 3+ levels (the combinatorial explosion of flat labels becomes unmanageable) |
| **Cardinality per level** | Fine with <100 total leaf classes | Shines when each level has manageable cardinality (e.g., 5 departments × 4 categories vs. 20 flat labels) |
| **Shared category names** | Treats `Mens/Clothing` and `Womens/Clothing` as unrelated | Naturally handles overlap — the department classifier routes first, then a specialized category model only chooses among that department's options |
| **Maintenance** | One model to retrain | Multiple models, but each is simpler and can be updated independently |

**The critical risk: cascading errors.** If the Department classifier routes a "smart watch" to `Mens` instead of `Electronics`, the downstream Category classifier is forced to choose from `Clothing`, `Shoes`, `Accessories`, or `Outerwear` — it never even sees `Mobile Phones` or `Computers` as options. The top-level mistake is unrecoverable.

Below we adapt three of our approaches — Embeddings, Centroids, and Gradient Boosting — to work hierarchically. (Levenshtein is omitted since it already struggled with flat classification.)

## Hierarchical Setup: Department and Category Centroids

For the local classifier approaches, we need centroids at two levels:
1. **Department centroids** — the average embedding of all products in each department
2. **Per-department category centroids** — the average embedding of products in each category, grouped by department

We also embed all new products upfront so we don't re-call the embedding API for each approach.

In [25]:
# Department-level centroids
dept_centroids = {}
for dept, group in catalog_df.groupby('department'):
    dept_centroids[dept] = np.mean(catalog_embeddings[group.index.tolist()], axis=0)

# Per-department category centroids
dept_category_centroids = {}  # {department: {category: centroid_vector}}
for dept in HIERARCHY:
    dept_category_centroids[dept] = {}
    dept_data = catalog_df[catalog_df['department'] == dept]
    for cat, cat_group in dept_data.groupby('category'):
        dept_category_centroids[dept][cat] = np.mean(catalog_embeddings[cat_group.index.tolist()], axis=0)

# Embed all new products once (reuse across hierarchical approaches)
print('Embedding new products...')
new_embed_response = client.models.embed_content(
    model=EMBEDDING_MODEL,
    contents=new_df['search_text'].tolist(),
    config=types.EmbedContentConfig(task_type='CLASSIFICATION'),
)
new_embeddings = np.array([emb.values for emb in new_embed_response.embeddings])

print(f'Department centroids: {len(dept_centroids)}')
print(f'Per-department category centroid groups: {sum(len(v) for v in dept_category_centroids.values())}')
print(f'New product embeddings: {new_embeddings.shape}')

Embedding new products...
Department centroids: 5
Per-department category centroid groups: 21
New product embeddings: (8, 3072)


---
## Approach 5: Hierarchical Embeddings (Top-Down Nearest Neighbor)

The hierarchical version of Approach 2. Instead of finding the single closest catalog item across the entire catalog, we:

1. **Step 1:** Find the closest catalog item *per department* and pick the department with the highest similarity
2. **Step 2:** Within that department, find the closest catalog item and inherit its category

This naturally handles the shared category name problem — "Clothing" under "Mens" only competes with other Mens products, not with "Clothing" under "Womens."

In [26]:
def classify_hierarchical_embeddings(new_vec, catalog_df, catalog_vecs):
    """Two-step classification: department first, then category within department."""
    # Step 1: Find the best department by nearest neighbor
    best_dept_score = -1
    best_dept = None
    for dept, group in catalog_df.groupby('department'):
        group_vecs = catalog_vecs[group.index.tolist()]
        similarities = cosine_similarity(new_vec.reshape(1, -1), group_vecs)[0]
        dept_best = similarities.max()
        if dept_best > best_dept_score:
            best_dept_score = dept_best
            best_dept = dept

    # Step 2: Within the predicted department, find the best category
    dept_data = catalog_df[catalog_df['department'] == best_dept]
    dept_vecs = catalog_vecs[dept_data.index.tolist()]
    similarities = cosine_similarity(new_vec.reshape(1, -1), dept_vecs)[0]
    best_idx = np.argmax(similarities)
    best_match = dept_data.iloc[best_idx]
    return best_dept, best_match['category'], best_dept_score, best_match['name']

print('--- APPROACH 5: HIERARCHICAL EMBEDDINGS (TOP-DOWN NN) ---\n')

hier_emb_results = []
for i, (_, new_prod) in enumerate(new_df.iterrows()):
    dept, cat, score, matched = classify_hierarchical_embeddings(
        new_embeddings[i], catalog_df, catalog_embeddings,
    )
    hier_emb_results.append({'sku': new_prod['sku'], 'department': dept, 'category': cat, 'score': round(score, 3)})
    print(f'{new_prod["name"]}')
    print(f'  -> {dept} / {cat}')
    print(f'  -> Dept match score: {score:.3f}, category via: "{matched}"\n')

--- APPROACH 5: HIERARCHICAL EMBEDDINGS (TOP-DOWN NN) ---

Classic Indigo Legwear
  -> Mens / Clothing
  -> Dept match score: 0.920, category via: "501 Original Jeans"

Glimmering Stilettos
  -> Womens / Shoes
  -> Dept match score: 0.847, category via: "Arizona Sandals"

Quantum View Panel
  -> Electronics / Televisions
  -> Dept match score: 0.873, category via: "BRAVIA XR 4K TV"

Rapid Brew System
  -> Home & Kitchen / Small Appliances
  -> Dept match score: 0.896, category via: "K-Classic Coffee Maker"

Aura Fitness Tracker
  -> Womens / Accessories
  -> Dept match score: 0.834, category via: "Charm Bracelet"

Dino Code-bot
  -> Children / Toys
  -> Dept match score: 0.830, category via: "Wooden Play Kitchen"

Summit Trail Jacket
  -> Womens / Outerwear
  -> Dept match score: 0.922, category via: "Down Sweater Jacket"

Align Sculpt Tights
  -> Womens / Activewear
  -> Dept match score: 0.916, category via: "Align Leggings"



---
## Approach 6: Hierarchical Centroids (Top-Down Category Averages)

The hierarchical version of Approach 3. We compare against department-level centroids first, then against category centroids within the predicted department.

This is the fastest hierarchical approach at scale — in a real catalog with 50 departments and 10 categories each, you'd compare against 50 department centroids then 10 category centroids, instead of 500 flat centroids or millions of individual products.

In [27]:
def classify_hierarchical_centroids(new_vec, dept_cents, dept_cat_cents):
    """Two-step centroid classification: department centroid first, then category centroid."""
    new_vec_2d = new_vec.reshape(1, -1)

    # Step 1: Best department centroid
    best_dept_score = -1
    best_dept = None
    for dept, centroid in dept_cents.items():
        score = cosine_similarity(new_vec_2d, [centroid])[0][0]
        if score > best_dept_score:
            best_dept_score = score
            best_dept = dept

    # Step 2: Best category centroid within that department
    best_cat_score = -1
    best_cat = None
    for cat, centroid in dept_cat_cents[best_dept].items():
        score = cosine_similarity(new_vec_2d, [centroid])[0][0]
        if score > best_cat_score:
            best_cat_score = score
            best_cat = cat

    return best_dept, best_cat, best_dept_score, best_cat_score

print('--- APPROACH 6: HIERARCHICAL CENTROIDS (TOP-DOWN) ---\n')

hier_cen_results = []
for i, (_, new_prod) in enumerate(new_df.iterrows()):
    dept, cat, dept_score, cat_score = classify_hierarchical_centroids(
        new_embeddings[i], dept_centroids, dept_category_centroids,
    )
    hier_cen_results.append({'sku': new_prod['sku'], 'department': dept, 'category': cat, 'score': round(cat_score, 3)})
    print(f'{new_prod["name"]}')
    print(f'  -> {dept} / {cat}')
    print(f'  -> Dept score: {dept_score:.3f}, Category score: {cat_score:.3f}\n')

--- APPROACH 6: HIERARCHICAL CENTROIDS (TOP-DOWN) ---

Classic Indigo Legwear
  -> Mens / Clothing
  -> Dept score: 0.916, Category score: 0.918

Glimmering Stilettos
  -> Womens / Shoes
  -> Dept score: 0.883, Category score: 0.863

Quantum View Panel
  -> Electronics / Televisions
  -> Dept score: 0.892, Category score: 0.886

Rapid Brew System
  -> Home & Kitchen / Small Appliances
  -> Dept score: 0.903, Category score: 0.913

Aura Fitness Tracker
  -> Electronics / Audio
  -> Dept score: 0.851, Category score: 0.834

Dino Code-bot
  -> Children / Toys
  -> Dept score: 0.853, Category score: 0.852

Summit Trail Jacket
  -> Womens / Outerwear
  -> Dept score: 0.906, Category score: 0.921

Align Sculpt Tights
  -> Womens / Activewear
  -> Dept score: 0.908, Category score: 0.931



---
## Approach 7: Hierarchical Gradient Boosting (Cascading Classifiers)

The hierarchical version of Approach 4. We train **separate models** at each level:
- **Level 1:** A single GBC that predicts Department (5 classes)
- **Level 2:** One GBC *per department* that predicts Category within that department (3-5 classes each)

This is the approach used by massive e-commerce taxonomies (Amazon, Walmart) where predicting 50,000+ leaf nodes flat is impractical. Each model is simpler and can be trained, evaluated, and updated independently.

**Note:** With our small catalog (~40 products, ~8 per department), each per-department classifier has very few training samples. In production you'd want hundreds of products per category at minimum.

In [28]:
# Level 1: Train a department classifier
dept_clf = GradientBoostingClassifier(n_estimators=50, learning_rate=0.1, max_depth=3, random_state=42)
dept_clf.fit(catalog_embeddings, catalog_df['department'].values)
print(f'Level 1 (Department): trained on {len(catalog_embeddings)} samples, {len(dept_clf.classes_)} classes')
print(f'  Classes: {list(dept_clf.classes_)}')

# Level 2: Train one category classifier per department
cat_clfs = {}
for dept in HIERARCHY:
    dept_mask = catalog_df['department'] == dept
    dept_X = catalog_embeddings[dept_mask.values]
    dept_y = catalog_df.loc[dept_mask, 'category'].values

    cat_clf = GradientBoostingClassifier(n_estimators=50, learning_rate=0.1, max_depth=3, random_state=42)
    cat_clf.fit(dept_X, dept_y)
    cat_clfs[dept] = cat_clf
    print(f'Level 2 ({dept}): trained on {len(dept_X)} samples, {len(cat_clf.classes_)} classes → {list(cat_clf.classes_)}')

Level 1 (Department): trained on 40 samples, 5 classes
  Classes: ['Children', 'Electronics', 'Home & Kitchen', 'Mens', 'Womens']
Level 2 (Mens): trained on 8 samples, 4 classes → ['Accessories', 'Clothing', 'Outerwear', 'Shoes']
Level 2 (Womens): trained on 10 samples, 5 classes → ['Accessories', 'Activewear', 'Clothing', 'Outerwear', 'Shoes']
Level 2 (Children): trained on 8 samples, 4 classes → ['Clothing', 'School Supplies', 'Shoes', 'Toys']
Level 2 (Electronics): trained on 8 samples, 4 classes → ['Audio', 'Computers', 'Mobile Phones', 'Televisions']
Level 2 (Home & Kitchen): trained on 6 samples, 4 classes → ['Cleaning', 'Cookware', 'Furniture', 'Small Appliances']


In [29]:
def classify_hierarchical_gbc(new_vec, dept_classifier, cat_classifiers):
    """Cascading GBC: predict department, then predict category within that department."""
    new_vec_2d = new_vec.reshape(1, -1)

    # Step 1: Predict department
    pred_dept = dept_classifier.predict(new_vec_2d)[0]
    dept_conf = np.max(dept_classifier.predict_proba(new_vec_2d))

    # Step 2: Predict category within department
    cat_clf = cat_classifiers[pred_dept]
    pred_cat = cat_clf.predict(new_vec_2d)[0]
    cat_conf = np.max(cat_clf.predict_proba(new_vec_2d))

    return pred_dept, pred_cat, dept_conf, cat_conf

print('--- APPROACH 7: HIERARCHICAL GRADIENT BOOSTING (CASCADING) ---\n')

hier_gbc_results = []
for i, (_, new_prod) in enumerate(new_df.iterrows()):
    dept, cat, dept_conf, cat_conf = classify_hierarchical_gbc(
        new_embeddings[i], dept_clf, cat_clfs,
    )
    hier_gbc_results.append({'sku': new_prod['sku'], 'department': dept, 'category': cat, 'score': round(cat_conf, 3)})
    print(f'{new_prod["name"]}')
    print(f'  -> {dept} / {cat}')
    print(f'  -> Dept confidence: {dept_conf:.3f}, Category confidence: {cat_conf:.3f}\n')

--- APPROACH 7: HIERARCHICAL GRADIENT BOOSTING (CASCADING) ---

Classic Indigo Legwear
  -> Womens / Clothing
  -> Dept confidence: 0.416, Category confidence: 0.319

Glimmering Stilettos
  -> Home & Kitchen / Cookware
  -> Dept confidence: 0.765, Category confidence: 0.429

Quantum View Panel
  -> Home & Kitchen / Cookware
  -> Dept confidence: 0.441, Category confidence: 0.442

Rapid Brew System
  -> Home & Kitchen / Small Appliances
  -> Dept confidence: 0.259, Category confidence: 0.930

Aura Fitness Tracker
  -> Children / Shoes
  -> Dept confidence: 0.504, Category confidence: 0.494

Dino Code-bot
  -> Children / Toys
  -> Dept confidence: 1.000, Category confidence: 0.433

Summit Trail Jacket
  -> Womens / Outerwear
  -> Dept confidence: 0.510, Category confidence: 0.690

Align Sculpt Tights
  -> Electronics / Audio
  -> Dept confidence: 0.336, Category confidence: 0.349



---
## Local Classifiers: Compare Hierarchical Approaches

Let's compare the three hierarchical approaches against each other, and see where the top-down routing differs from the flat approaches above.

In [30]:
# Hierarchical comparison
hier_comparison_rows = []
for i, (_, new_prod) in enumerate(new_df.iterrows()):
    he = hier_emb_results[i]
    hc = hier_cen_results[i]
    hg = hier_gbc_results[i]
    # Flat best (embeddings) for reference
    fe = embedding_results[i]

    row = {
        'Product': f'{new_prod["brand"]} {new_prod["name"]}',
        'Flat (Emb)': f'{fe["department"]} / {fe["category"]}',
        'Hier Emb': f'{he["department"]} / {he["category"]}',
        'Hier Centroid': f'{hc["department"]} / {hc["category"]}',
        'Hier GBC': f'{hg["department"]} / {hg["category"]}',
    }

    # Flag where hierarchical disagrees with flat
    flat_pred = row['Flat (Emb)']
    hier_preds = {row['Hier Emb'], row['Hier Centroid'], row['Hier GBC']}
    row['Flat vs Hier'] = 'agree' if flat_pred in hier_preds else 'DIFFER'

    hier_comparison_rows.append(row)

hier_comparison_df = pd.DataFrame(hier_comparison_rows)
hier_comparison_df

,Product,Flat (Emb),Hier Emb,Hier Centroid,Hier GBC,Flat vs Hier
0,Levi's Classic Indigo Legwear,Mens / Clothing,Mens / Clothing,Mens / Clothing,Womens / Clothing,agree
1,Steve Madden Glimmering Stilettos,Womens / Shoes,Womens / Shoes,Womens / Shoes,Home & Kitchen / Cookware,agree
2,Samsung Quantum View Panel,Electronics / Televisions,Electronics / Televisions,Electronics / Televisions,Home & Kitchen / Cookware,agree
3,Keurig Rapid Brew System,Home & Kitchen / Small Appliances,Home & Kitchen / Small Appliances,Home & Kitchen / Small Appliances,Home & Kitchen / Small Appliances,agree
4,Fitbit Aura Fitness Tracker,Womens / Accessories,Womens / Accessories,Electronics / Audio,Children / Shoes,agree
5,VTech Dino Code-bot,Children / Toys,Children / Toys,Children / Toys,Children / Toys,agree
6,The North Face Summit Trail Jacket,Womens / Outerwear,Womens / Outerwear,Womens / Outerwear,Womens / Outerwear,agree
7,Lululemon Align Sculpt Tights,Womens / Activewear,Womens / Activewear,Womens / Activewear,Electronics / Audio,agree


---
# Part 3: Multi-Task Learning (Deep Learning)

Multi-task learning offers the **best of both worlds**: a single model that understands the hierarchy without cascading errors.

**How it works:** A Keras neural network takes the embedding vector as input and splits into **two output heads** — one for Department, one for Category. The hidden layers are shared, so the network learns representations that are useful for *both* tasks simultaneously. The Department head informs the Category head through shared weights, without the rigid routing of local classifiers.

**When to use multi-task learning:**

| Factor | Local Classifiers | Multi-Task Learning |
|--------|-------------------|---------------------|
| **Cascading errors** | Fatal — wrong department means wrong category | No cascading — both heads predict independently from shared features |
| **Shared categories** | Handled by routing (never sees wrong department's categories) | Learned implicitly — shared layers discover that "Clothing" under different departments shares structure |
| **Model count** | 1 + N (one per department) models to deploy | Single model with two outputs |
| **Data efficiency** | Each sub-model only sees its department's data | All data trains the shared layers — better with limited labels |
| **Hierarchy depth** | Scales linearly (one model per level per node) | Add an output head per level — shared layers stay the same |

**The trade-off:** Requires a deep learning framework (Keras/TensorFlow or PyTorch) and more careful tuning. For simple taxonomies, the added complexity may not justify the accuracy gain. But as hierarchies get deeper and categories overlap more, multi-task learning scales more gracefully than maintaining dozens of cascading classifiers.

## Approach 8: Multi-Task Learning — Independent Heads

In [36]:
import os
os.environ['KERAS_BACKEND'] = 'tensorflow'

import keras
from sklearn.preprocessing import LabelEncoder

print(f'Keras version: {keras.__version__}')
print(f'Keras backend: {keras.backend.backend()}')

# Encode labels as integers for Keras
dept_encoder = LabelEncoder()
cat_encoder = LabelEncoder()

y_dept = dept_encoder.fit_transform(catalog_df['department'].values)
y_cat = cat_encoder.fit_transform(catalog_df['category'].values)

n_departments = len(dept_encoder.classes_)
n_categories = len(cat_encoder.classes_)
embedding_dim = catalog_embeddings.shape[1]

print(f'\nDepartment classes ({n_departments}): {list(dept_encoder.classes_)}')
print(f'Category classes ({n_categories}): {list(cat_encoder.classes_)}')
print(f'Embedding dimension: {embedding_dim}')

Keras version: 3.13.2
Keras backend: tensorflow

Department classes (5): ['Children', 'Electronics', 'Home & Kitchen', 'Mens', 'Womens']
Category classes (15): ['Accessories', 'Activewear', 'Audio', 'Cleaning', 'Clothing', 'Computers', 'Cookware', 'Furniture', 'Mobile Phones', 'Outerwear', 'School Supplies', 'Shoes', 'Small Appliances', 'Televisions', 'Toys']
Embedding dimension: 3072


The architecture is straightforward:

```
Embedding Input (3072-dim)
        │
   Dense(128, relu)    ← shared representation
        │
   Dense(64, relu)     ← shared representation
       ╱ ╲
      ╱   ╲
Dense(n_dept)  Dense(n_cat)   ← separate output heads
 softmax        softmax
```

Both heads share the same hidden layers. The loss from both heads backpropagates through the shared layers during training, teaching them to extract features useful for *both* department and category prediction simultaneously.

In [37]:
# Build multi-output Keras model
input_layer = keras.Input(shape=(embedding_dim,), name='embedding_input')

# Shared hidden layers
x = keras.layers.Dense(128, activation='relu', name='shared_1')(input_layer)
x = keras.layers.Dropout(0.3)(x)
x = keras.layers.Dense(64, activation='relu', name='shared_2')(x)
x = keras.layers.Dropout(0.3)(x)

# Output head 1: Department
dept_output = keras.layers.Dense(n_departments, activation='softmax', name='department')(x)

# Output head 2: Category
cat_output = keras.layers.Dense(n_categories, activation='softmax', name='category')(x)

model = keras.Model(inputs=input_layer, outputs=[dept_output, cat_output])

model.compile(
    optimizer='adam',
    loss={
        'department': 'sparse_categorical_crossentropy',
        'category': 'sparse_categorical_crossentropy',
    },
    metrics={'department': 'accuracy', 'category': 'accuracy'},
)

model.summary()

Model: "functional_1"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ embedding_input     │ (None, 3072)      │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ shared_1 (Dense)    │ (None, 128)       │    393,344 │ embedding_input[… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 128)       │          0 │ shared_1[0][0]    │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ shared_2 (Dense)    │ (None, 64)        │      8,256 │ dropout_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_3 (Dropout) │ (None, 64)        │          0 │ shared_2[0][0]    │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ department (Dense)  │ (None, 5)         │        325 │ dropout_3[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ category (Dense)    │ (None, 15)        │        975 │ dropout_3[0][0]   │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 402,900 (1.54 MB)

 Trainable params: 402,900 (1.54 MB)

 Non-trainable params: 0 (0.00 B)

In [38]:
# Train the multi-output model
history = model.fit(
    catalog_embeddings,
    {'department': y_dept, 'category': y_cat},
    epochs=200,
    batch_size=8,
    verbose=0,
)

print(f'Training complete ({len(history.epoch)} epochs)')
print(f'  Final department accuracy: {history.history["department_accuracy"][-1]:.3f}')
print(f'  Final category accuracy:   {history.history["category_accuracy"][-1]:.3f}')

Training complete (200 epochs)
  Final department accuracy: 1.000
  Final category accuracy:   1.000


In [39]:
# Predict on new products
dept_preds, cat_preds = model.predict(new_embeddings, verbose=0)

print('--- APPROACH 8: MULTI-TASK LEARNING (KERAS) ---\n')

mtl_results = []
for i, (_, new_prod) in enumerate(new_df.iterrows()):
    dept_idx = np.argmax(dept_preds[i])
    cat_idx = np.argmax(cat_preds[i])
    dept = dept_encoder.inverse_transform([dept_idx])[0]
    cat = cat_encoder.inverse_transform([cat_idx])[0]
    dept_conf = dept_preds[i][dept_idx]
    cat_conf = cat_preds[i][cat_idx]

    mtl_results.append({'sku': new_prod['sku'], 'department': dept, 'category': cat,
                        'dept_conf': round(float(dept_conf), 3), 'cat_conf': round(float(cat_conf), 3)})
    print(f'{new_prod["name"]}')
    print(f'  -> {dept} / {cat}')
    print(f'  -> Dept confidence: {dept_conf:.3f}, Category confidence: {cat_conf:.3f}\n')

--- APPROACH 8: MULTI-TASK LEARNING (KERAS) ---

Classic Indigo Legwear
  -> Mens / Clothing
  -> Dept confidence: 0.973, Category confidence: 0.994

Glimmering Stilettos
  -> Womens / Shoes
  -> Dept confidence: 0.998, Category confidence: 0.946

Quantum View Panel
  -> Electronics / Televisions
  -> Dept confidence: 1.000, Category confidence: 0.796

Rapid Brew System
  -> Home & Kitchen / Small Appliances
  -> Dept confidence: 0.999, Category confidence: 0.969

Aura Fitness Tracker
  -> Electronics / Activewear
  -> Dept confidence: 0.637, Category confidence: 0.390

Dino Code-bot
  -> Children / Toys
  -> Dept confidence: 0.996, Category confidence: 0.828

Summit Trail Jacket
  -> Mens / Outerwear
  -> Dept confidence: 0.576, Category confidence: 0.998

Align Sculpt Tights
  -> Womens / Activewear
  -> Dept confidence: 1.000, Category confidence: 0.994



---
## Approach 9: Multi-Task Learning — Conditional Category Head

The independent-heads model above can predict **Electronics / Activewear** — a department/category pair that doesn't exist in our hierarchy. This happens because the two softmax heads know nothing about each other. Three approaches can address this:

1. **Post-prediction masking** — After inference, zero out category probabilities that are invalid for the predicted department and re-normalize. Simple, no model change needed, but the model never *learns* the constraints — you're just patching its output.

2. **Conditional architecture** — Feed the department prediction *into* the category head so it learns to condition its category choice on the predicted department. The model discovers valid combinations from the training data naturally.

3. **Custom masking layer** — Bake a hard validity mask into the network itself (e.g., multiply category logits by a binary mask derived from the department prediction before the softmax). This *guarantees* invalid combinations are impossible, but adds architectural complexity.

Below we implement **approach 2** — it's the most natural for a neural network and teaches the category head to respect departmental boundaries without hard-coding rules.

### Conditional Architecture

```
Embedding Input (3072-dim)
        │
   Dense(128, relu)      ← shared representation
        │
   Dense(64, relu)       ← shared representation
       │ ╲
       │  Dense(n_dept)  ← department head
       │  softmax
       │ ╱
  Concatenate([shared, dept_pred])
        │
   Dense(32, relu)       ← category-specific layer
        │
   Dense(n_cat)          ← category head (conditioned on department)
    softmax
```

The key difference: the category head receives the 64-dim shared features **plus** the 5-dim department probability vector. During training, gradients flow through both paths, so the category head learns that when the department activations look like "Electronics", it should favor categories like Televisions and Computers — not Activewear.

In [40]:
# Build conditional multi-output model
input_layer = keras.Input(shape=(embedding_dim,), name='embedding_input')

# Shared hidden layers (same as before)
x = keras.layers.Dense(128, activation='relu', name='shared_1')(input_layer)
x = keras.layers.Dropout(0.3)(x)
x = keras.layers.Dense(64, activation='relu', name='shared_2')(x)
x = keras.layers.Dropout(0.3)(x)

# Department head (same as before)
dept_output = keras.layers.Dense(n_departments, activation='softmax', name='department')(x)

# Category head: conditioned on department prediction
cat_input = keras.layers.Concatenate(name='cat_conditioning')([x, dept_output])
cat_hidden = keras.layers.Dense(32, activation='relu', name='cat_hidden')(cat_input)
cat_output = keras.layers.Dense(n_categories, activation='softmax', name='category')(cat_hidden)

cond_model = keras.Model(inputs=input_layer, outputs=[dept_output, cat_output])

cond_model.compile(
    optimizer='adam',
    loss={
        'department': 'sparse_categorical_crossentropy',
        'category': 'sparse_categorical_crossentropy',
    },
    metrics={'department': 'accuracy', 'category': 'accuracy'},
)

cond_model.summary()

Model: "functional_2"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ embedding_input     │ (None, 3072)      │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ shared_1 (Dense)    │ (None, 128)       │    393,344 │ embedding_input[… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_4 (Dropout) │ (None, 128)       │          0 │ shared_1[0][0]    │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ shared_2 (Dense)    │ (None, 64)        │      8,256 │ dropout_4[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_5 (Dropout) │ (None, 64)        │          0 │ shared_2[0][0]    │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ department (Dense)  │ (None, 5)         │        325 │ dropout_5[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ cat_conditioning    │ (None, 69)        │          0 │ dropout_5[0][0],  │
│ (Concatenate)       │                   │            │ department[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ cat_hidden (Dense)  │ (None, 32)        │      2,240 │ cat_conditioning… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ category (Dense)    │ (None, 15)        │        495 │ cat_hidden[0][0]  │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 404,660 (1.54 MB)

 Trainable params: 404,660 (1.54 MB)

 Non-trainable params: 0 (0.00 B)

In [41]:
# Train the conditional model
cond_history = cond_model.fit(
    catalog_embeddings,
    {'department': y_dept, 'category': y_cat},
    epochs=200,
    batch_size=8,
    verbose=0,
)

print(f'Training complete ({len(cond_history.epoch)} epochs)')
print(f'  Final department accuracy: {cond_history.history["department_accuracy"][-1]:.3f}')
print(f'  Final category accuracy:   {cond_history.history["category_accuracy"][-1]:.3f}')

Training complete (200 epochs)
  Final department accuracy: 0.975
  Final category accuracy:   1.000


In [42]:
# Predict with the conditional model
cond_dept_preds, cond_cat_preds = cond_model.predict(new_embeddings, verbose=0)

# Build validity lookup for checking predictions
valid_combos = set()
for dept, cats in HIERARCHY.items():
    for cat in cats:
        valid_combos.add((dept, cat))

print('--- APPROACH 9: CONDITIONAL MULTI-TASK LEARNING (KERAS) ---\n')

cond_mtl_results = []
for i, (_, new_prod) in enumerate(new_df.iterrows()):
    dept_idx = np.argmax(cond_dept_preds[i])
    cat_idx = np.argmax(cond_cat_preds[i])
    dept = dept_encoder.inverse_transform([dept_idx])[0]
    cat = cat_encoder.inverse_transform([cat_idx])[0]
    dept_conf = cond_dept_preds[i][dept_idx]
    cat_conf = cond_cat_preds[i][cat_idx]

    is_valid = (dept, cat) in valid_combos
    validity_flag = '' if is_valid else '  ⚠ INVALID COMBINATION'

    cond_mtl_results.append({'sku': new_prod['sku'], 'department': dept, 'category': cat,
                             'dept_conf': round(float(dept_conf), 3), 'cat_conf': round(float(cat_conf), 3),
                             'valid': is_valid})
    print(f'{new_prod["name"]}')
    print(f'  -> {dept} / {cat}{validity_flag}')
    print(f'  -> Dept confidence: {dept_conf:.3f}, Category confidence: {cat_conf:.3f}\n')

# Summary
n_valid = sum(1 for r in cond_mtl_results if r['valid'])
print(f'Valid combinations: {n_valid}/{len(cond_mtl_results)}')

--- APPROACH 9: CONDITIONAL MULTI-TASK LEARNING (KERAS) ---

Classic Indigo Legwear
  -> Mens / Clothing
  -> Dept confidence: 0.982, Category confidence: 0.997

Glimmering Stilettos
  -> Womens / Shoes
  -> Dept confidence: 0.989, Category confidence: 0.998

Quantum View Panel
  -> Electronics / Televisions
  -> Dept confidence: 0.999, Category confidence: 0.988

Rapid Brew System
  -> Home & Kitchen / Small Appliances
  -> Dept confidence: 0.989, Category confidence: 0.956

Aura Fitness Tracker
  -> Womens / Activewear
  -> Dept confidence: 0.878, Category confidence: 0.627

Dino Code-bot
  -> Children / Toys
  -> Dept confidence: 0.998, Category confidence: 0.977

Summit Trail Jacket
  -> Womens / Outerwear
  -> Dept confidence: 0.534, Category confidence: 0.998

Align Sculpt Tights
  -> Womens / Activewear
  -> Dept confidence: 1.000, Category confidence: 0.996

Valid combinations: 8/8


---
# Final Comparison: All Paradigms

Let's put all nine approaches side-by-side across the three paradigms — flat, local/hierarchical, and multi-task — to see the full picture.

In [43]:
# Full comparison across all paradigms
full_comparison_rows = []
for i, (_, new_prod) in enumerate(new_df.iterrows()):
    row = {
        'Product': f'{new_prod["brand"]} {new_prod["name"]}',
        # Flat approaches (best two)
        'Flat Emb': f'{embedding_results[i]["department"]} / {embedding_results[i]["category"]}',
        'Flat GBC': f'{gbc_results[i]["department"]} / {gbc_results[i]["category"]}',
        # Hierarchical approaches
        'Hier Emb': f'{hier_emb_results[i]["department"]} / {hier_emb_results[i]["category"]}',
        'Hier GBC': f'{hier_gbc_results[i]["department"]} / {hier_gbc_results[i]["category"]}',
        # Multi-task (independent heads)
        'MTL': f'{mtl_results[i]["department"]} / {mtl_results[i]["category"]}',
        # Multi-task (conditional)
        'MTL Cond': f'{cond_mtl_results[i]["department"]} / {cond_mtl_results[i]["category"]}',
    }

    # Count unique predictions across all approaches
    all_preds = [row['Flat Emb'], row['Flat GBC'], row['Hier Emb'], row['Hier GBC'], row['MTL'], row['MTL Cond']]
    row['Consensus'] = f'{6 - len(set(all_preds)) + 1}/6'

    full_comparison_rows.append(row)

full_comparison_df = pd.DataFrame(full_comparison_rows)
full_comparison_df

,Product,Flat Emb,Flat GBC,Hier Emb,Hier GBC,MTL,MTL Cond,Consensus
0,Levi's Classic Indigo Legwear,Mens / Clothing,Womens / Clothing,Mens / Clothing,Womens / Clothing,Mens / Clothing,Mens / Clothing,5/6
1,Steve Madden Glimmering Stilettos,Womens / Shoes,Womens / Shoes,Womens / Shoes,Home & Kitchen / Cookware,Womens / Shoes,Womens / Shoes,5/6
2,Samsung Quantum View Panel,Electronics / Televisions,Children / School Supplies,Electronics / Televisions,Home & Kitchen / Cookware,Electronics / Televisions,Electronics / Televisions,4/6
3,Keurig Rapid Brew System,Home & Kitchen / Small Appliances,Home & Kitchen / Small Appliances,Home & Kitchen / Small Appliances,Home & Kitchen / Small Appliances,Home & Kitchen / Small Appliances,Home & Kitchen / Small Appliances,6/6
4,Fitbit Aura Fitness Tracker,Womens / Accessories,Children / Shoes,Womens / Accessories,Children / Shoes,Electronics / Activewear,Womens / Activewear,3/6
5,VTech Dino Code-bot,Children / Toys,Children / Toys,Children / Toys,Children / Toys,Children / Toys,Children / Toys,6/6
6,The North Face Summit Trail Jacket,Womens / Outerwear,Mens / Outerwear,Womens / Outerwear,Womens / Outerwear,Mens / Outerwear,Womens / Outerwear,5/6
7,Lululemon Align Sculpt Tights,Womens / Activewear,Womens / Activewear,Womens / Activewear,Electronics / Audio,Womens / Activewear,Womens / Activewear,5/6


### Observations from This Run

> **Note:** The catalog and new products are generated by Gemini on each run, so your results will differ. The observations below are from one specific execution and illustrate the *types* of failures each paradigm produces — your run will surface its own interesting cases.

**Approach 1 (Levenshtein) struggles with semantic gaps.** Fuzzy matching placed *"Glimmering Stilettos"* into **Children / School Supplies** (matched to "SuperBreak Backpack" at 56/100) and routed *"Dino Code-bot"* to **Womens / Accessories** (matched to "Charm Bracelet" at 55/100). These are character-overlap accidents — the approach has no understanding of what these products actually are.

**Approach 4 (Flat GBC) overfits with limited data.** With only ~2 training samples per class, the Gradient Boosting Classifier made confident but wrong predictions — placing *"Quantum View Panel"* (a TV) into **Children / School Supplies** with 0.731 confidence, and *"Aura Fitness Tracker"* into **Children / Shoes** with 0.657 confidence. The sklearn warning about class count exceeding 50% of samples is well-earned.

**Approach 7 (Hierarchical GBC) demonstrates cascading errors perfectly.** When the department classifier routed *"Glimmering Stilettos"* to **Home & Kitchen**, the category classifier was forced to choose among Cleaning, Cookware, Furniture, and Small Appliances — landing on **Cookware**. The same cascade sent *"Align Sculpt Tights"* (yoga pants) to **Electronics / Audio**. Once the top-level prediction is wrong, the category prediction is guaranteed to be wrong too.

**Approach 8 (Independent-heads MTL) produces invalid combinations.** The model predicted *"Aura Fitness Tracker"* as **Electronics / Activewear** — but "Activewear" isn't a valid category under Electronics in our hierarchy. Because the two output heads are independent, they can each be individually reasonable (Electronics makes sense, Activewear makes sense) while producing a combination that doesn't exist in the taxonomy.

**Approach 9 (Conditional MTL) fixes the invalid combination problem.** By feeding the department prediction into the category head, the model learns which categories are valid for each department. The conditional model produced **8/8 valid** department/category pairs. For the *"Aura Fitness Tracker"*, it predicted **Womens / Activewear** — a valid combination — with department confidence 0.878 and category confidence 0.627. The lower confidence correctly signals this is still an ambiguous product.

**Approaches 2, 5, and 9 (embedding-based) were the most consistent.** Flat Embeddings (Approach 2) and Hierarchical Embeddings (Approach 5) agreed on every product in this run, and the Conditional MTL (Approach 9) largely aligned with them. All three handled synonym-heavy descriptions well ("legwear" → Clothing, "display panel" → Televisions). The *"Aura Fitness Tracker"* remained the hardest product — a genuinely ambiguous item that could reasonably be Electronics, Accessories, or Activewear depending on the taxonomy's intent.

---
## Choosing a Paradigm: Decision Guide

| Scenario | Recommended Paradigm | Why |
|----------|---------------------|-----|
| **Shallow taxonomy** (2 levels, <100 leaf classes) | **Flat Classification** | Simplest to build, train, and deploy. No structural overhead needed. |
| **Deep taxonomy** (3+ levels, 1000s of leaf paths) | **Local Classifiers** | Flat classification can't scale to 50,000 leaf labels. Top-down routing keeps each model's search space manageable. |
| **High category overlap** ("Clothing" in Mens, Womens, Children) | **Local Classifiers** or **Multi-Task** | Flat treats `Mens/Clothing` and `Womens/Clothing` as unrelated. Both hierarchical paradigms understand the structure. |
| **Limited labeled data** | **Multi-Task Learning** | Shared layers train on all data, not just one department's subset. Better data efficiency than cascading classifiers. |
| **Zero cascading error tolerance** | **Multi-Task Learning** | Independent output heads — a wrong department prediction doesn't doom the category prediction. |
| **Easiest to maintain in production** | **Flat Classification** | One model, one deployment. Local classifiers require N+1 models; multi-task needs a DL stack. |

### Summary for Data Engineers

**Start with Flat Classification** — it's the easiest to build, deploy, and maintain. Use embedding-based approaches (centroids or nearest neighbor) for the best accuracy-to-complexity ratio.

**Upgrade to Local Classifiers** when your taxonomy grows beyond what a single flat model can handle — typically at 3+ hierarchy levels or 500+ leaf categories. Be prepared to invest in robust department-level classification to minimize cascading errors.

**Upgrade to Multi-Task Learning** when you need the accuracy benefits of hierarchical awareness without the cascading error risk, or when you have limited labeled data per category and need the shared-layer data efficiency. The Keras model we built here is a starting point — production systems often add hierarchy-aware loss functions that penalize cross-department errors more heavily than within-department errors.

---
## Further Directions

- **LLM-based classification:** Use Gemini directly as a zero-shot or few-shot classifier with the hierarchy as context. Handles nuanced reasoning (e.g., "kids' learning tablet" → Children/Toys, not Electronics/Computers) without any training data.
- **Hierarchy-aware loss functions:** Custom losses that penalize predicting the wrong department more than predicting the wrong category within the right department.
- **Multi-node placement:** Products that belong in multiple categories (threshold-based similarity, multi-label classifiers, or LLM with multi-output schema).
- **Hybrid approaches:** Embedding retrieval for candidate generation + LLM for final classification. Fast approach for high-confidence items, LLM escalation for ambiguous ones.
- **Production deployment:** BigQuery ML for at-scale training, Vertex AI Feature Store for embedding caching, batch pipelines for vendor onboarding, real-time endpoints for individual submissions, feedback loops from human corrections.